drivers/linstorvolumemanager.py

```diff
@@ -196,17 +202,88 @@ def _get_controller_uri():
         # Not found, maybe we are trying to create the SR...
         pass
 
-def get_controller_uri():
-    retries = 0
-    while True:
-        uri = _get_controller_uri()
+
+@contextlib.contextmanager
+def shared_reader(path):
+    with open(path, 'r') as f:
+        lock = flock.ReadLock(f.fileno())
+        lock.lock()
+        try:
+            yield f
+        finally:
+            lock.unlock()
+
+
+@contextlib.contextmanager
+def excl_writer(path):
+    with open(path, 'r+') as f:
+        lock = flock.WriteLock(f.fileno())
+        lock.lock()
+        try:
+            yield f
+        finally:
+            lock.unlock()
+
+
+def get_cached_controller_uri(ctx=None):
+    try:
+        with ctx if ctx is not None else shared_reader(CONTROLLER_CACHE_PATH) as f:
+            return f.read().strip()
+    except FileNotFoundError:
+        pass
+    except Exception as e:
+        util.SMlog('Unable to read controller URI cache file at `{}` : {}'.format(CONTROLLER_CACHE_PATH,e))
+
+
+def delete_controller_uri_cache(ctx=None):
+    try:
+        with ctx if ctx is not None else excl_writer(CONTROLLER_CACHE_PATH) as f:
+            f.seek(0)
+            f.truncate()
+    except FileNotFoundError:
+        pass
+    except Exception as e:
+        util.SMlog('Unable to delete uri cache file at `{}` : {}'.format(CONTROLLER_CACHE_PATH, e))
+
+
+def write_controller_uri_cache(uri, ctx=None):
+    try:
+        if not os.path.exists(CONTROLLER_CACHE_DIRECTORY):
+            os.makedirs(CONTROLLER_CACHE_DIRECTORY)
+            os.chmod(CONTROLLER_CACHE_DIRECTORY, 0o700)
+        with ctx if ctx is not None else excl_writer(CONTROLLER_CACHE_PATH) as f:
```

Comment:
```suggestion
        with ctx if ctx else excl_writer(CONTROLLER_CACHE_PATH) as f:
```
Same here

Reply:
fixed in [4b663cb](https://github.com/xcp-ng/sm/pull/83/commits/4b663cb99b7162d1e218094611a60c7c616f63d0)